src/coreclr/jit/lower.cpp

```diff
@@ -5447,6 +5449,60 @@ GenTree* Lowering::LowerAdd(GenTreeOp* node)
             JITDUMP("Remove [%06u], [%06u]\n", op2->gtTreeID, node->gtTreeID);
             return next;
         }
+#ifdef TARGET_ARM64
+        if (BlockRange().TryGetUse(node, &use))
+        {
+            if (comp->opts.OptimizationEnabled() && varTypeIsIntegral(node) && !node->gtOverflow() &&
```

Comment:
Is there a reason why this optimization is done during lower instead of morph?

Reply:
Yes. There are a couple of reasons:

- There are no `GT_MADD`/`GT_MSUB` HIR ops that can be used in the morph phase.
- Re-arranging the nodes in morph, like what we are doing in lowering, change what CSEs get picked up.
    - Originally in this PR, I did the re-arrangement in morph. When I looked at the codegen, there were several places that regressed. I wish we had a morph-like phase after all CSE phases are complete instead of having to do the work in lowering.